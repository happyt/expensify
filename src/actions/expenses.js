import uuid from 'uuid'

// ADD ENTRY
export const addExpense = (
    {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0
        } = {}
) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
    }
})

// REMOVE ENTRY
export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    id
})

// EDIT ENTRY
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})