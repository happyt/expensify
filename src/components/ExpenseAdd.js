import React from 'react'

const ExpenseAdd = () => (
    <div>Create entry</div>
)
export default ExpenseAdd