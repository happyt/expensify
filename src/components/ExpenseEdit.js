import React from 'react'

const ExpenseEdit = () => (
    <div>Edit entry</div>
)
export default ExpenseEdit