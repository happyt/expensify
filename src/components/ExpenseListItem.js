import React from 'react'

const ExpenseListItem = ({ description, amount, createdAt}) => (
    <div>
        <div>{description}</div>
        <p>{amount} - {createdAt}</p>
    </div>
)
export default ExpenseListItem

