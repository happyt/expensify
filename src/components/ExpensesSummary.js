import React from 'react'
import numeral from 'numeral'
import { connect } from 'react-redux'
import selectExpenses from '../selectors/expenses'
import getExpensesTotal from '../selectors/expenses-total'

const ExpenseSummary = (props) => (
    <div>
        <p>Viewing {props.expenses.length} expense{props.expenses.length > 1? 's totalling ' : ' value '}  
        {numeral(getExpensesTotal(props.expenses)/100).format('Â£0,0.00')}
         
         </p>
    </div>
)
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseSummary)
