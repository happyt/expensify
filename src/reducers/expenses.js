
const expensesReducerDefaultState = []

export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ]
        case 'REMOVE_EXPENSE':
            //          console.log('Id: ' + action.id)
            return state.filter((exp) => {
                return (action.id !== exp.id)
            })
        case 'EDIT_EXPENSE':
            return state.map((exp) => {
                if (exp.id === action.id) {
                    return {
                        ...exp,
                        ...action.updates
                    }
                } else {
                    return exp
                }
            })
        case 'SET_EXPENSES':
        return action.expenses
        default:
            return state
    }
}