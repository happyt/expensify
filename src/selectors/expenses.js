import moment from 'moment'

export default (expenses, {text, sortBy, startDate, endDate}) => {
    const textSearch = text.toLowerCase()
   return expenses.filter((exp) => {
    const createdAtMoment = moment(exp.createdAt)
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
    // const startDateMatch = typeof startDate !== 'number' || exp.createdAt >= startDate 
    // const endDateMatch = typeof endDate !== 'number' || exp.createdAt <= endDate 
 
       const itemDesc = exp.description && exp.description.toLowerCase(); 
       const textMatch = itemDesc.includes(textSearch) 

       return startDateMatch && endDateMatch && textMatch
   }).sort((a, b) => {
       if (sortBy === 'date') {
           return a.createdAt < b.createdAt ? 1 : 0
       } else if (sortBy === 'amount') {
           return a.amount < b.amount ? 1 : 0            
       }
   })
}