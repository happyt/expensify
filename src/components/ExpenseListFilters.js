import React from 'react'
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates'
import { setTextFilter, sortBy, setStartDate, setEndDate } from '../actions/filters'

class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange =(calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }

    onSortChange = (e) => {
        this.props.sortBy(e.target.value)
    }

    render() {

        return (
            <div>
                <input type='text' value={this.props.filters.text}
                    onChange={this.onTextChange} />
                <select value={this.props.filters.sortBy}
                    onChange={this.onSortChange} >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
        filters: state.filters
    })

const mapDispatchToProps = (dispatch) => ({
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        sortBy: (value) => dispatch(sortBy(value)),
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    })

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)