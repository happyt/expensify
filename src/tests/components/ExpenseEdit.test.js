import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseForm } from '../../components/ExpenseForm';
import { ExpenseEdit } from '../../components/ExpenseEdit';
import expenses from '../fixtures/expenses';

let editExpense, removeExpense, history, wrapper;

beforeEach(() => {
    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(
        <ExpenseEdit 
            editExpense={editExpense} 
            removeExpense={removeExpense} 
            history={history}
            expense={expenses[2]}
         />)
  })

test('should render ExpenseEdit page', () => {
    expect(wrapper).toMatchSnapshot()
});

test('should handle edit expense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2])
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2])
});

test('should handle remove expense', () => {
    wrapper.find('button').simulate('click')
    expect(removeExpense).toHaveBeenLastCalledWith(expenses[2].id)
});
