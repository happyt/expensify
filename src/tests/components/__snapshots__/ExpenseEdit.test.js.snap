// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseEdit page 1`] = `
<div>
  <p>
    Expense Form
  </p>
  <ExpenseForm
    expense={
      Object {
        "amount": 4500,
        "createdAt": 345600000,
        "description": "Credit card",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
